feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/navigation/ForYouNavigation.kt

```diff
@@ -19,28 +19,24 @@ package com.google.samples.apps.nowinandroid.feature.foryou.navigation
 import androidx.navigation.NavController
 import androidx.navigation.NavGraphBuilder
 import androidx.navigation.NavOptions
-import androidx.navigation.NavType
 import androidx.navigation.compose.composable
-import androidx.navigation.navArgument
 import androidx.navigation.navDeepLink
 import com.google.samples.apps.nowinandroid.feature.foryou.ForYouRoute
+import kotlinx.serialization.Serializable
 
 const val LINKED_NEWS_RESOURCE_ID = "linkedNewsResourceId"
-const val FOR_YOU_ROUTE = "for_you_route/{$LINKED_NEWS_RESOURCE_ID}"
 private const val DEEP_LINK_URI_PATTERN =
     "https://www.nowinandroid.apps.samples.google.com/foryou/{$LINKED_NEWS_RESOURCE_ID}"
 
-fun NavController.navigateToForYou(navOptions: NavOptions) = navigate(FOR_YOU_ROUTE, navOptions)
+@Serializable data class ForYouDestination(val linkedNewsResourceId: String? = null)
+
+fun NavController.navigateToForYou(navOptions: NavOptions) = navigate(route = ForYouDestination(), navOptions)
 
 fun NavGraphBuilder.forYouScreen(onTopicClick: (String) -> Unit) {
-    composable(
-        route = FOR_YOU_ROUTE,
+    composable<ForYouDestination>(
         deepLinks = listOf(
             navDeepLink { uriPattern = DEEP_LINK_URI_PATTERN },
```

Comment:
Yes, another good spot. The name of the argument placeholder in the deeplink currently has to be identical to the property in the `ForYouDestination`. That duplication needs to be removed. 

The new API offers the following: 

```
navDeepLink<ForYouDestination>(basePath = DEEP_LINK_URI_PATTERN),
```

However, I'm not sure how the mapping works between the serializable data class property `ForYouDestination.linkedNewsResourceId` and the deeplink URI format (currently `https://www.nowinandroid.apps.samples.google.com/foryou/{linkedNewsResourceId}`). 

In our specific case it doesn't actually matter, as long as it works, because the only place it's used is when creating a PendingIntent for a notification (in `SystemTrayNotifier`) so there's no external dependency on the deeplink URI pattern. However, I can imagine that in other apps it'd be difficult to change the deeplink URI pattern so we should figure out how to do this migration.  

I will ask the Jetpack team to clarify guidance here. Incidentally, the AndroidX sample has the same issue: https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:navigation/navigation-compose/integration-tests/navigation-demos/src/main/java/androidx/navigation/compose/demos/NavByDeepLinkDemo.kt;l=54 `userId` is both a string in the URI, and a property of `Dashboard`.

Reply:
Kotlin Serialization [supports](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/basic-serialization.md#serial-field-names) having a different serialized value from the property name with `@SerialName`, which I would assume is the mechanism we could make use of here if they didn't match.

Reply:
The [kdoc here](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:navigation/navigation-common/src/main/java/androidx/navigation/NavDeepLink.kt;l=469) explains how the final route is generated for deeplinks. 

In this case, since `linkedNewsResourceId` has a default value (null), it would be appended as a query parameter. 

You'd probably want `DEEP_LINK_URI_PATTERN` to be `https://www.nowinandroid.apps.samples.google.com/foryou/`, and you would end up with this deeplink
`https://www.nowinandroid.apps.samples.google.com/foryou?linkedNewsResourceId={linkedNewsResourceId}`

On a side note, please feel free to let me know how we might improve the kdocs for this. I want to make sure people understand what the final deeplink looks like.

Reply:
Nulls get encoded to a string literal null, so it will be
`https://www.nowinandroid.apps.samples.google.com/foryou?linkedNewsResourceId=null`

Reply:
Thanks @claraf3. According to the docs, in order to keep our existing format, which uses _path parameters_, we have to have a destination argument _without_ a default value. 

Changing the destination's constructor to change the deeplink pattern results in some duplicated code at the call site. For example, we now have 2 calls to `ForYouDestination(linkedNewsResourceId=null)` which is an obvious case for a default value, but if I add a default value, the argument is represented as a query parameter which breaks our original format. 

Any advice here?

Reply:
What are the consequences of breaking the original format? If it is absolutely crucial to keep it, a workaround is to add a deeplink to the NavDestination with a new class where the `linkedNewsResourceId` does not have a default value.

For example
```
@Serializable
data class ForYouDestinationAlternative(val linkedNewsResourceId: String?)
```

This way you can navigate to the destination with either deeplink:
`https://www.nowinandroid.apps.samples.google.com/foryou?linkedNewsResourceId=null`
`https://www.nowinandroid.apps.samples.google.com/foryou/null`

Reply:
> What are the consequences of breaking the original format?

None for this app, but could be severe for another app wanting to migrate to use these APIs.

I've updated the PR to include this workaround just to demonstrate that it's possible.

Reply:
After online discussion, since the workaround was removed, I think we need to update the corresponding logic in `SystemTrayNotifier` to use the query parameter instead of the path parameter.

Reply:
I've refactored this so that the `linkedNewsResourceId` isn't part of the route and is supplied as a path parameter as it was previously. More details here: https://github.com/android/nowinandroid/blob/dt/nav-safe-args/feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/navigation/ForYouNavigation.kt#L36